### PR TITLE
Add brpc serialization support.

### PR DESCRIPTION
**Pull request by gongweibao**
Fix part of https://github.com/PaddlePaddle/Paddle/issues/10804

### CODE REVIEW

File: paddle/fluid/operators/detail/brpc_server.cc [hunk 1]
```diff
@@ -13,31 +13,34 @@
...
 
 class BRPCServiceImpl : public SendRecvService {
  public:
   explicit BRPCServiceImpl(const HandlerMap& rpc_call_map)
       : request_send_h_(nullptr),
         request_get_h_(nullptr),
         request_prefetch_h_(nullptr) {
-    auto it = rpc_call_map.find(paddle::operators::detail::kRequestSend);
+    VLOG(3) << "BRPCServiceImpl size: " << rpc_call_map.size();
+    auto it = rpc_call_map.find(detail::kRequestSend);
```
Comment (Tuvie) [inline paddle/fluid/operators/detail/brpc_server.cc]: In fact, you can also use brpc to do the method mapping.

File: paddle/fluid/operators/detail/brpc_variable_response.h [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+class BRPCVariableResponse : public VariableResponse {
+ public:
+  BRPCVariableResponse(const framework::Scope* scope,
+                       const platform::DeviceContext* dev_ctx,
+                       bool create_scope = false)
+      : VariableResponse(scope, dev_ctx, create_scope) {}
+
+  virtual ~BRPCVariableResponse() {}
+
+  // parse attchment from iobuf
```
Comment (Tuvie) [inline paddle/fluid/operators/detail/brpc_variable_response.h]: attachment

File: paddle/fluid/operators/detail/brpc_server.cc (hunk 1 above)
Comment (gongweibao) [reply paddle/fluid/operators/detail/brpc_server.cc]: Please give me an example.

File: paddle/fluid/operators/detail/brpc_variable_response.h (hunk 2 above)
Comment (gongweibao) [reply paddle/fluid/operators/detail/brpc_variable_response.h]: Done. Thanks!

File: paddle/fluid/operators/detail/brpc_server.cc (hunk 1 above)
Comment (Tuvie) [reply paddle/fluid/operators/detail/brpc_server.cc]: For each service you define in the "proto" file, you can add more than one method. Please refer to https://github.com/Tuvie/brpc/blob/master/example/http_c%2B%2B/http.proto

File: paddle/fluid/operators/detail/macros.h [hunk 3]
```diff
@@ -14,14 +14,22 @@
```
Comment (seiriosPlus) [inline paddle/fluid/operators/detail/macros.h]: macros.h文件是只给rpc相关使用，还是其他OP也能使用， 如果其他OP也能用，不需要RPC相关的OP引入RPC相关的头文件感觉不合适

File: paddle/fluid/operators/distributed/brpc_client.cc [hunk 4]
```diff
@@ -26,18 +27,27 @@ DEFINE_int32(max_retry, 3, "Max retries(not including the first RPC)");
...
   std::unique_ptr<brpc::Controller> cntl_guard(cntl);
   std::unique_ptr<sendrecv::VoidMessage> response_guard(response);
 
+  // this channel can be used by other now.
+  ch_ptr->Push(ch_ctx);
+  cls->DecreaseReqCount();
+
   if (cntl->Failed()) {
-    LOG(WARNING) << "Fail to send EchoRequest, " << cntl->ErrorText();
+    LOG(FATAL) << "Fail to send SendVar: " << var_h.name
```
Comment (seiriosPlus) [inline paddle/fluid/operators/distributed/brpc_client.cc]: LOG(FATAL) -> VLOG

File: paddle/fluid/framework/lod_tensor.cc [hunk 5]
```diff
@@ -68,7 +68,7 @@ std::ostream &operator<<(std::ostream &os, const LoDTensor &t) {
   // only print first ten elements
   int64_t size = t.numel() < 10 ? t.numel() : 10;
   for (int64_t i = 0; i < size; ++i) {
-    if (t.type().hash_code() == typeid(float).hash_code()) {
+    if (t.type().hash_code() == typeid(float).hash_code()) {  // NOLINT
```
Comment (typhoonzero) [inline paddle/fluid/framework/lod_tensor.cc]: Can put these fix in another PR to reduce this PR's size?

File: paddle/fluid/operators/distributed/brpc_client.h [hunk 6]
```diff
@@ -55,33 +56,62 @@ class BRPCClient : public RPCClient {
...
   void Proceed();
   ChannelQueuePtr GetChannel(const std::string& ep);
 
+  void AsyncSendComplete(const std::string& ep,
+                         int64_t time_out = FLAGS_rpc_deadline);
+
+  void AsyncSendMessage(const std::string& ep, const std::string& message,
+                        int64_t time_out);
+
+  friend void HandleSendResponse(brpc::Controller* cntl,
```
Comment (typhoonzero) [inline paddle/fluid/operators/distributed/brpc_client.h:124]: Curious why need to be "friend"?

File: paddle/fluid/framework/lod_tensor.cc (hunk 5 above)
Comment (gongweibao) [reply paddle/fluid/framework/lod_tensor.cc]: Ye.Will do it!

File: paddle/fluid/operators/distributed/brpc_client.h (hunk 6 above)
Comment (gongweibao) [reply paddle/fluid/operators/distributed/brpc_client.h:124]: This function will call this class private functions.

File: benchmark/fluid/fluid_benchmark.py [hunk 7]
```diff
@@ -81,9 +81,11 @@ def dist_transpile(trainer_id, args, train_prog, startup_prog):
     # the role, should be either PSERVER or TRAINER
     training_role = os.getenv("PADDLE_TRAINING_ROLE")
 
-    config = distribute_transpiler.DistributeTranspilerConfig()
+    config = fluid.DistributeTranspilerConfig()
     config.slice_var_up = not args.no_split_var
+    config.min_block_size = 1048576
```
Comment (panyx0718) [inline benchmark/fluid/fluid_benchmark.py:86]: why not use the default for benchmark?

File: cmake/external/brpc.cmake [hunk 8]
```diff
@@ -31,14 +31,15 @@ SET(BRPC_LIBRARIES "${BRPC_INSTALL_DIR}/lib/libbrpc.a" CACHE FILEPATH "brpc libr
...
+set(prefix_path "${THIRD_PARTY_PATH}/install/gflags|${THIRD_PARTY_PATH}/install/leveldb|${THIRD_PARTY_PATH}/install/snappy|${THIRD_PARTY_PATH}/install/gtest|${THIRD_PARTY_PATH}/install/protobuf|${THIRD_PARTY_PATH}/install/zlib|${THIRD_PARTY_PATH}/install/glog")
 
 # If minimal .a is need, you can set  WITH_DEBUG_SYMBOLS=OFF
 ExternalProject_Add(
     extern_brpc
     ${EXTERNAL_PROJECT_LOG_ARGS}
+    # TODO(gongwb): change to de newst repo when they changed.
     GIT_REPOSITORY  "https://github.com/gongweibao/brpc"
-    GIT_TAG         "7dc04defad1fd4173aae170c3fcbde131b65155a"
+    GIT_TAG         "master"
```
Comment (panyx0718) [inline cmake/external/brpc.cmake]: are you sure? This might not be stable.

File: cmake/external/gtest.cmake [hunk 9]
```diff
@@ -12,8 +12,12 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-IF(WITH_TESTING)
-    ENABLE_TESTING()
+#FIXME:(gongwb) Move brpc's gtest dependency.
+IF(WITH_TESTING OR (WITH_DISTRIBUTE AND NOT WITH_GRPC))
```
Comment (panyx0718) [inline cmake/external/gtest.cmake:16]: seems not safe. fix it now?

File: paddle/fluid/framework/CMakeLists.txt [hunk 10]
```diff
@@ -144,7 +144,11 @@ cc_library(feed_fetch_method SRCS feed_fetch_method.cc DEPS lod_tensor scope glo
 cc_library(naive_executor SRCS naive_executor.cc DEPS op_registry device_context scope framework_proto glog lod_rank_table feed_fetch_method graph_to_program_pass)
 
 if(WITH_DISTRIBUTE)
-  cc_library(executor SRCS executor.cc DEPS op_registry device_context scope framework_proto glog lod_rank_table feed_fetch_method sendrecvop_grpc cares grpc++_unsecure grpc_unsecure gpr graph_to_program_pass)
+  if(WITH_GRPC)
+      cc_library(executor SRCS executor.cc DEPS op_registry device_context scope framework_proto glog lod_rank_table feed_fetch_method ${GLOB_DISTRIBUTE_DEPS} gpr graph_to_program_pass)
+  else(WITH_GRPC)
+    cc_library(executor SRCS executor.cc DEPS op_registry device_context scope framework_proto glog lod_rank_table feed_fetch_method ${GLOB_DISTRIBUTE_DEPS})
```
Comment (panyx0718) [inline paddle/fluid/framework/CMakeLists.txt]: no need to be different?

File: paddle/fluid/inference/analysis/CMakeLists.txt [hunk 11]
```diff
@@ -34,7 +34,7 @@ function (inference_analysis_test TARGET)
         endif()
         cc_test(${TARGET}
                 SRCS "${analysis_test_SRCS}"
-                DEPS analysis pass ${GLOB_PASS_LIB} ${analysis_test_EXTRA_DEPS}
+                DEPS analysis pass ${GLOB_PASS_LIB} ${analysis_test_EXTRA_DEPS} paddle_fluid ${GLOB_DISTRIBUTE_DEPS}
```
Comment (panyx0718) [inline paddle/fluid/inference/analysis/CMakeLists.txt]: why would inference need distributed_deps?

File: paddle/fluid/inference/api/CMakeLists.txt [hunk 12]
```diff
@@ -34,7 +34,7 @@ function(inference_api_test TARGET_NAME)
         set(PYTHON_TESTS_DIR ${PADDLE_BINARY_DIR}/python/paddle/fluid/tests)
         cc_test(${TARGET_NAME}
                 SRCS ${inference_test_SRC}
-                DEPS "${inference_deps}"
+                DEPS "${inference_deps}" ${GLOB_DISTRIBUTE_DEPS}
```
Comment (panyx0718) [inline paddle/fluid/inference/api/CMakeLists.txt]: why inference need distributed deps?

File: paddle/fluid/operators/distributed/brpc_client.cc [hunk 13]
```diff
@@ -14,128 +14,243 @@
...
   std::unique_ptr<sendrecv::VariableMessage> response_guard(response);
 
+  // this channel can be used other now.
+  ch_ptr->Push(ch_ctx);
+
   if (cntl->Failed()) {
-    LOG(WARNING) << "Fail to send EchoRequest, " << cntl->ErrorText();
+    LOG(FATAL) << "Fail to send SendVar: " << var_h->name()
+               << ", error text: " << cntl->ErrorText();
+    cls->DecreaseReqCount();
```
Comment (panyx0718) [inline paddle/fluid/operators/distributed/brpc_client.cc:136]: why not move this out?

File: paddle/fluid/operators/distributed/brpc_client.cc [hunk 14]
```diff
@@ -14,128 +14,243 @@
...
-  LOG(INFO) << "Received response from " << cntl->remote_side()
-            << " latency=" << cntl->latency_us() << "us";
 
-  // framework::Variable* outvar = nullptr;
-  // DeserializeFromByteBuffer(ret_msg, *var_h.ctx, var_h.scope, &outvar);
+  VLOG(4) << "Received SendResponse from: " << cntl->remote_side()
+          << ", varname: " << var_h->name()
+          << ", latency: " << cntl->latency_us() << "us";
+
+  framework::Variable* outvar = nullptr;
```
Comment (panyx0718) [inline paddle/fluid/operators/distributed/brpc_client.cc:145]: Is this used somewhere?

File: paddle/fluid/operators/distributed/brpc_client.h (hunk 6 above)
Comment (panyx0718) [reply paddle/fluid/operators/distributed/brpc_client.h:124]: should they just be private method of this class?

File: paddle/fluid/operators/distributed/brpc_client.h [hunk 15]
```diff
@@ -53,33 +55,78 @@ class BRPCClient : public RPCClient {
...
+
+  friend void HandleFetchBarrierResponse(brpc::Controller* cntl,
+                                         sendrecv::VariableMessage* response,
+                                         VarHandlePtr var_h,
+                                         ChannelQueuePtr ch_ptr,
+                                         ChannelContextPtr ch_ctx,
+                                         BRPCClient* cls);
+  void DecreaseReqCount() {
+    req_count_--;
+    if (req_count_.load() <= 0) {
```
Comment (panyx0718) [inline paddle/fluid/operators/distributed/brpc_client.h]: `if (--req_count_ <= 0)`?

File: paddle/fluid/operators/distributed/brpc_rdma_pool.cc [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+  if (old != nullptr) {
+    if (data != old) {
+      PADDLE_ENFORCE(false, "var:%s data:%ld != %ld", varname, data, old);
+    }
+    VLOG(7) << "Find on rdma:" << varname << " data:" << data
+            << " data_size:" << data_size;
+    return;
+  }
+
+  pthread_rwlock_wrlock(&access_);
```
Comment (panyx0718) [inline paddle/fluid/operators/distributed/brpc_rdma_pool.cc]: It seems you only need to protect line68?

File: paddle/fluid/operators/distributed/brpc_rdma_pool.cc [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+    LOG(FATAL) << "register " << varname << " data:" << data
+               << " data_size:" << data_size << " error";
+  }
+
+  pthread_rwlock_unlock(&access_);
+
+  VLOG(4) << "register on rdma:" << varname << " data:" << data
+          << " data_size:" << data_size;
+
+  return;
```
Comment (panyx0718) [inline paddle/fluid/operators/distributed/brpc_rdma_pool.cc]: remove?

File: paddle/fluid/operators/distributed/brpc_sendrecvop_utils.cc [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+
+    // GPU data is copied to CPU buffer when sending,
+    // free the buffer when possible.
+    platform::CUDAPinnedPlace cuda_pinned;
+    memory::Free(cuda_pinned, p);
+#endif
+  }
+
+  static void Append(butil::IOBuf* iobuf, int k, const char* v, int64_t vlen) {
+    iobuf->append(reinterpret_cast<char*>(&k), 4);
```
Comment (panyx0718) [inline paddle/fluid/operators/distributed/brpc_sendrecvop_utils.cc:35]: change int k to int32_t k?

File: CMakeLists.txt [hunk 19]
```diff
@@ -66,7 +66,7 @@ option(WITH_FAST_BUNDLE_TEST    "Bundle tests that can be run in a single proces
 option(WITH_CONTRIB     "Compile the third-party contributation"        OFF)
 option(REPLACE_ENFORCE_GLOG "Replace PADDLE_ENFORCE with glog/CHECK for better debug." OFF)
 option(WITH_ANAKIN      "Compile with Anakin library"                   OFF)
-option(WITH_GRPC     "Use grpc as the default rpc framework"            ${WITH_DISTRIBUTE})
+option(WITH_GRPC     "Use grpc as the default rpc framework"            OFF)
```
Comment (Yancey0623) [inline CMakeLists.txt]: Keep the optional argument for GRPC.

File: paddle/fluid/operators/distributed/brpc_server.cc [hunk 20]
```diff
@@ -51,46 +55,93 @@ class BRPCServiceImpl : public SendRecvService {
...
+
+    if (outvar) {
+      distributed::SerializeToIOBuf(varname, outvar, *request_get_h_->dev_ctx(),
+                                    response, &cntl->response_attachment(), "",
+                                    false);
+    }
+  }
+
+  /*
+  static void HandlePrefetchRequest(distributed::VariableResponse* resp){
```
Comment (Yancey0623) [inline paddle/fluid/operators/distributed/brpc_server.cc]: Delete the commented code?

File: benchmark/fluid/fluid_benchmark.py (hunk 7 above)
Comment (gongweibao) [reply benchmark/fluid/fluid_benchmark.py:86]: Large block size can get better performance.

File: cmake/external/brpc.cmake (hunk 8 above)
Comment (gongweibao) [reply cmake/external/brpc.cmake]: I'll change it when they merge into brpc main branch.

File: cmake/external/gtest.cmake (hunk 9 above)
Comment (gongweibao) [reply cmake/external/gtest.cmake:16]: Gtest is needed by brpc now.

File: paddle/fluid/framework/CMakeLists.txt (hunk 10 above)
Comment (gongweibao) [reply paddle/fluid/framework/CMakeLists.txt]: Thanks! Done!

File: paddle/fluid/operators/distributed/brpc_client.cc (hunk 13 above)
Comment (gongweibao) [reply paddle/fluid/operators/distributed/brpc_client.cc:136]: Request count==0是代表请求都已经返回。这个需要在处理完成以后才能做。

File: paddle/fluid/operators/distributed/brpc_client.cc (hunk 14 above)
Comment (gongweibao) [reply paddle/fluid/operators/distributed/brpc_client.cc:145]: Line 137.

File: paddle/fluid/operators/distributed/brpc_client.h (hunk 15 above)
Comment (gongweibao) [reply paddle/fluid/operators/distributed/brpc_client.h]: Thanks! Done!

File: paddle/fluid/operators/distributed/brpc_rdma_pool.cc (hunk 16 above)
Comment (gongweibao) [reply paddle/fluid/operators/distributed/brpc_rdma_pool.cc]: Thanks! Done!

File: paddle/fluid/operators/distributed/brpc_rdma_pool.cc (hunk 17 above)
Comment (gongweibao) [reply paddle/fluid/operators/distributed/brpc_rdma_pool.cc]: Thanks! Done!

File: paddle/fluid/operators/distributed/brpc_sendrecvop_utils.cc (hunk 18 above)
Comment (gongweibao) [reply paddle/fluid/operators/distributed/brpc_sendrecvop_utils.cc:35]: https://github.com/brpc/brpc/blob/master/src/butil/iobuf.h#L214

File: paddle/fluid/operators/distributed/brpc_server.cc (hunk 20 above)
Comment (gongweibao) [reply paddle/fluid/operators/distributed/brpc_server.cc]: Thanks! Done!

File: CMakeLists.txt (hunk 19 above)
Comment (gongweibao) [reply CMakeLists.txt]: It's optional now!

Comment (gongweibao) [reply CMakeLists.txt]: Thanks!Done!

File: paddle/fluid/inference/analysis/CMakeLists.txt (hunk 11 above)
Comment (gongweibao) [reply paddle/fluid/inference/analysis/CMakeLists.txt]: Done.

File: cmake/external/brpc.cmake (hunk 8 above)
Comment (jacquesqiao) [reply cmake/external/brpc.cmake]: why not use one commit of brpc repo?

File: cmake/external/brpc.cmake [hunk 21]
```diff
@@ -65,5 +68,6 @@ ADD_LIBRARY(brpc STATIC IMPORTED GLOBAL)
 SET_PROPERTY(TARGET brpc PROPERTY IMPORTED_LOCATION ${BRPC_LIBRARIES})
 ADD_DEPENDENCIES(brpc extern_brpc)
 
+add_definitions(-DBRPC_WITH_GLOG)
```
Comment (jacquesqiao) [inline cmake/external/brpc.cmake:71]: What is this used for?

File: paddle/fluid/framework/CMakeLists.txt [hunk 22]
```diff
@@ -165,9 +165,15 @@ cc_library(variable_helper SRCS variable_helper.cc DEPS lod_tensor)
 cc_library(naive_executor SRCS naive_executor.cc DEPS op_registry device_context scope framework_proto glog lod_rank_table feed_fetch_method graph_to_program_pass variable_helper)
 
 if(WITH_DISTRIBUTE)
-  cc_library(executor SRCS executor.cc DEPS op_registry device_context scope framework_proto glog lod_rank_table feed_fetch_method sendrecvop_grpc cares grpc++_unsecure grpc_unsecure gpr graph_to_program_pass variable_helper)
+    if(WITH_GRPC)
+  cc_library(executor SRCS executor.cc DEPS op_registry device_context scope framework_proto glog lod_rank_table feed_fetch_method sendrecvop_grpc cares grpc++_unsecure grpc_unsecure gpr ${GLOB_DISTRIBUTE_DEPS} graph_to_program_pass variable_helper)
```
Comment (jacquesqiao) [inline paddle/fluid/framework/CMakeLists.txt]: This if/else have many duplicated things, maybe SEND_RECV_OP_RPC is better.
```cmake
if (WITH_GRPC)
  set(SEND_RECV_OP_RPC sendrecvop_grpc)
else(WITH_BRPC)
  set(SEND_RECV_OP_RPCsendrecvop_brpc )
endif()
  
````

File: cmake/external/brpc.cmake (hunk 21 above)
Comment (gongweibao) [reply cmake/external/brpc.cmake:71]: Brpc use this macro to distinguish glog with brpc's own log.
So when we use brpc headerfiles ,we should add this.

File: paddle/fluid/framework/CMakeLists.txt (hunk 22 above)
Comment (gongweibao) [reply paddle/fluid/framework/CMakeLists.txt]: Done. Thanks.

File: paddle/fluid/operators/distributed/brpc_sendrecvop_utils.cc [hunk 23]
```diff
@@ -0,0 +1,196 @@
...
+                                void (*destroy)(void*), void* user_data) {
+    VLOG(7) << "AppendTCPZeroCopy "
+            << " k:" << k
+            << " data:" << static_cast<void*>(const_cast<char*>(v))
+            << " data_size:" << vlen << " in_cuda_pinned:" << in_cuda_pinned;
+
+    iobuf->append(reinterpret_cast<char*>(&k), 4);
+    iobuf->append(reinterpret_cast<char*>(&vlen), 8);
+
+    // FIXME(gongwb): use append_zerocopy
```
Comment (typhoonzero) [inline paddle/fluid/operators/distributed/brpc_sendrecvop_utils.cc:51]: why cannot use this? need brpc support?

File: paddle/fluid/operators/distributed/brpc_sendrecvop_utils.cc [hunk 24]
```diff
@@ -0,0 +1,196 @@
...
+#endif
+  } else {
+    PADDLE_THROW("Serialize does not support type: %s",
+                 typeid(var->Type()).name());
+  }
+
+  PADDLE_ENFORCE_NOT_NULL(payload);
+
+  // FIXME(gongwb): it seems that can use zero copy.
+  if (var_is_not_stable) {
```
Comment (typhoonzero) [inline paddle/fluid/operators/distributed/brpc_sendrecvop_utils.cc:150]: what `var_is_not_stable` means?

**Review by typhoonzero — DISMISSED** [review]
LGTM basically, but still need a way to check this through CI later.

File: paddle/fluid/operators/distributed/brpc_sendrecvop_utils.cc (hunk 23 above)
Comment (gongweibao) [reply paddle/fluid/operators/distributed/brpc_sendrecvop_utils.cc:51]: We changed tensor memory holder, and brpc didn't support destroy userdata but only support destroy the payload data.

File: paddle/fluid/operators/distributed/brpc_sendrecvop_utils.cc (hunk 24 above)
Comment (gongweibao) [reply paddle/fluid/operators/distributed/brpc_sendrecvop_utils.cc:150]: When a var is a temp variable, it's memory can be free out, so we can't append only a pointer but copy data.
Yuyang change it tensor data holder to sharedptr and then this not useful.I'll clean up next pr.

File: paddle/fluid/operators/distributed/brpc_client.h (hunk 6 above)
Comment (gongweibao) [reply paddle/fluid/operators/distributed/brpc_client.h:124]: We need a function point here and it can't support function object.

File: paddle/fluid/operators/distributed/brpc_server.cc [hunk 25]
```diff
@@ -13,84 +13,287 @@
...
-    auto it = rpc_call_map.find(paddle::operators::distributed::kRequestSend);
+  explicit BRPCServiceImpl(const HandlerMap& rpc_call_map,
+                           distributed::RPCServer* rpc_server)
+      : rpc_server_(rpc_server) {
+    VLOG(3) << "BRPCServiceImpl size: " << rpc_call_map.size();
+    auto it = rpc_call_map.find(distributed::kRequestSend);
     if (it != rpc_call_map.end()) {
       request_send_h_ = it->second;
+      send_threads_.reset(new paddle::framework::ThreadPool(
+          rpc_server_->GetThreadNum(distributed::kRequestSend)));
```
Comment (typhoonzero) [inline paddle/fluid/operators/distributed/brpc_server.cc:38]: can use `FLAGS_rpc_send_thread_num` etc

Comment (gongweibao) [reply paddle/fluid/operators/distributed/brpc_server.cc:38]: The caller will use `FLAGS_rpc_send_thread_num` to register this property and it use the arguments.

**Review by typhoonzero — APPROVED** [review]
